pgrx/src/pg_catalog.rs

```diff
@@ -0,0 +1,1212 @@
+                    Self: 'src,
+                {
+                    Self::new(datum.sans_lifetime().value() as c_char)
+                }
+            }
+
+            unsafe impl GetStruct<c_char> for [<$table:camel $column:camel>] {
+                unsafe fn get_struct(raw: *const c_char) -> Self {
+                    unsafe { Self::new(raw.read()) }
+                }
+            }
+        }
+    };
+}
+
+macro_rules! define_column {
+    {
+        $(#[$column_meta:meta])* column ($table:ident, $column:ident, $column_type:ty, $(#[$character_meta:meta])* character { $($(#[$variant_meta:meta])* ($variant:ident, $variant_value:literal))* } $($x:tt)*)
+    } => {
+        paste::paste! {
+            define_character! { $(#[$character_meta])* character ($table, $column) { $($(#[$variant_meta])* ($variant, $variant_value))* } }
+            define_column! { $(#[$column_meta])* column ($table, $column, $column_type $($x)*) }
+        }
+    };
+    {
+        $(#[$column_meta:meta])* column ($table:ident, $column:ident, $column_type:ty, get_struct)
+    } => {
+        paste::paste! {
+            impl [<$table:camel>]<'_> {
+                $(#[$column_meta])*
+                pub fn $column(&self) -> $column_type {
+                    unsafe {
+                        let start = self.inner.t_data.cast::<u8>();
+                        let offset = (*self.inner.t_data).t_hoff as usize;
+                        let p = start.add(offset).cast::<pgrx::pg_sys::[<FormData_ $table>]>();
+                        GetStruct::get_struct(std::ptr::addr_of!((*p).$column))
+                    }
+                }
+            }
+        }
+    };
+    {
+        $(#[$column_meta:meta])* column ($table:ident, $column:ident, $column_type:ty, get_attr)
+    } => {
+        paste::paste! {
+            impl [<$table:camel>]<'_> {
+                $(#[$column_meta])*
+                pub fn $column(&self) -> Option<$column_type> {
+                    self.get_attr::<$column_type>(pgrx::pg_sys::[<Anum_ $table _ $column>])
+                }
+            }
+        }
+    };
+    {
+        $(#[$column_meta:meta])* column ($table:ident, $column:ident, $column_type:ty, get_attr_notnull)
+    } => {
+        paste::paste! {
+            impl [<$table:camel>]<'_> {
+                $(#[$column_meta])*
+                pub fn $column(&self) -> $column_type {
+                    self.get_attr::<$column_type>(pgrx::pg_sys::[<Anum_ $table _ $column>]).unwrap()
+                }
+            }
+        }
+    };
+}
+
+macro_rules! define_catalog {
+    {
+        $(#[$table_meta:meta])* catalog ($table:ident) { $($(#[$column_meta:meta])* ($column:ident, $column_type:ty, $($x:tt)*))* }
+    } => {
+        paste::paste!{
+            #[doc = concat!("Safe wrapper for pg_catalog.", stringify!($table), ".")]
+            $(#[$table_meta])*
+            pub struct [<$table:camel>]<'a> {
+                inner: &'a pgrx::pg_sys::HeapTupleData,
```

Comment:
Normally I avoid holding references to variably-sized structs due to [UCG#256](https://github.com/rust-lang/unsafe-code-guidelines/issues/256) still being unresolved. I think it doesn't matter in this case.